src/org/opensolaris/opengrok/util/OptionParser.java

```diff
@@ -0,0 +1,701 @@
+                            ii--;
+                        }
+                    }
+                }
+                
+                // If there is no value setting for the 
+                // option by now, throw a hissy fit.
+                if (opt.value == null) {
+                    throw new ParseException("Option " + option + " requires a value.", ii);
+                }
+
+                // Only specific values allowed?
+                if (opt.allowedValues != null) {
+                    if (!opt.allowedValues.contains(opt.value)) {
+                        throw new ParseException(
+                           "'"+ opt.value + 
+                           "' is unknown value for option " + opt.names + 
+                           ". Must be one of " + opt.allowedValues, ii);
+                    }
+                }
+                
+                Object value = opt.value;
+                
+                // Should option argument match some pattern?
+                if (opt.valuePattern != null) {
+                    Matcher m = opt.valuePattern.matcher(opt.value);
+                    if (!m.matches()) {
+                        throw new ParseException(
+                           "Value '"+ opt.value + "' for option " + opt.names+opt.argument +
+                           "\n does not match pattern " + opt.valuePattern, ii);
+                    }
+                
+                // Handle special conversions of input 
+                // arguments before sending to action handler.
+                } else if (opt.valueType != null) {
+                    
+                    if (!converters.containsKey(opt.valueType)) {
+                        throw new ParseException(
+                            "No conversion handler for data type " + opt.valueType, ii);
+                    }
+                    
+                    try {
+                        DataParser data = converters.get(opt.valueType);
+                        value = data.converter.apply(opt.value);
+                        
+                    } catch (Exception e) {
+                        throw new ParseException(
+                           "Failed to parse ("+opt.value+") as value of " + opt.names, ii);
+                    }
+                }
+                
+                if (opt.action != null) {
+                    opt.action.accept(value); // 'do' assigned action
+                }
+                optind = ++ii;
+            }
+        }
+        
+        // Prepare to gather any remaining arguments
+        // to send back to calling program.
+        
+        String[] remainingArgs = null;
+        
+        if (optind == -1) {
+            remainingArgs = args;
+        } else if (optind < args.length) {
+            remainingArgs = Arrays.copyOfRange(args, optind, args.length);
+        }
+        
+        return remainingArgs;
+    }
+    
+    private String program() {
```

Comment:
I don't like this - surely there is better way how to get program name other than traversing call stack ?

Reply:
I was hoping for something already available as an interface in java, but web searching wasn't popping up anything easily relevant. Some of the solutions were downright scary looking. So I chose the least scary.  If you have pointers for a better way... let me know.

Reply:
maybe just specify the progname while constructing the parser..

Reply:
Note that this is only a fallback mechanism (default) when the programmer doesn't specify a prologue. So in a way, the program name can already be specified there.

Reply:
I would suggest the main class to be a parameter of a constructor as this class is not very useful fro any other use than with a main class. How about that?

Reply:
Or better part of this call opts.parse(); where the args is quite tight to the main class

Reply:
When someone wants to name the program, this is how its done currently:
    
    String program = "opengrok.jar";
    openGrok = OptionParser.Do(parser -> {
        parser.setPrologue(
	     String.format("\nUsage: %s [options] [subDir1 [...]]\n", program));

I don't see a need for anything else. I reiterate that this 'program'method is just an internal mechanism to generate a placeholder name when someone doesn't specify a proglogue. I am pretty sure most people will be setting the prologue value to their particular needs.

Reply:
better to have no fallback than hacky fallback I think

Reply:
Vlad, why do you think this code is too hacky (as in a bad sense)?  After I obtained this from the net, I did a step through the code during a debugging session to see what it was doing. It certainly makes sense that the bottom of the stack is going to be the entry point of the program. Look at the enclosed snapshot I took when I activated the Indexer program.

![image](https://user-images.githubusercontent.com/13596544/31673147-78027baa-b32c-11e7-8c55-3fbac5529729.png)

So you would rather just have the default for prologue be,    Usage: MyProgram [options]
than the name of the actual entry point?

Reply:
yes.

Reply:
As for the hackiness, I think traversing the stack which is behind-the-covers data structure to obtain program name is too much hassle. Given the `argv[0]` in Java is not the program name, it better be supplied by hand.